sdk/identity/azure-identity/inc/azure/identity/managed_identity_credential.hpp

```diff
@@ -42,6 +42,91 @@ namespace Azure { namespace Identity {
     std::string ToString() const { return m_resourceId; }
   };
 
+  /**
+   * @brief The type of managed identity identifier depending on how the managed identity is
+   * configured.
+   *
+   * @remark This can either be system assigned, or user assigned which corresponds to an identifier
+   * that represents either client ID, resource ID, or object ID, depending on how the managed
+   * identity is configuerd.
+   */
+  enum class ManagedIdentityIdType
+  {
+    SystemAssigned,
+    ClientId,
+    ObjectId,
+    ResourceId,
+  };
+
+  /**
+   * @brief The type of managed identity and its corresponding identifier.
+   *
+   * @remark This property holds the type and unique identifier of either a system or user-assigned
+   * managed identity.
+   */
+  class ManagedIdentityType final {
+  private:
+    ManagedIdentityIdType m_idType;
+    std::string m_id;
+
+  public:
+    /**
+     * @brief Constructs the type of managed identity.
+     *
+     * @remark This defaults to ManagedIdentityIdType::SystemAssigned.
+     */
+    explicit ManagedIdentityType() : m_idType(ManagedIdentityIdType::SystemAssigned) {}
+
+    /**
+     * @brief Constructs the type of managed identity.
+     *
+     * @param idType The type of the managed identity identifier.
+     * @param id The value of the managed identity identifier. This can be either a client ID,
+     * resource ID, or object ID.
+     *
+     * @remark For ManagedIdentityIdType::SystemAssigned, the id must be an empty string.
+     *
+     * @remark Make sure the type of ID matches the value of the ID. For example, the client
+     * ID and object ID are NOT interchangeable, even though they are both Uuid values.
+     */
+    explicit ManagedIdentityType(ManagedIdentityIdType idType, std::string id = std::string())
+        : m_idType(idType), m_id(id)
+    {
+      if (idType == ManagedIdentityIdType::SystemAssigned && !id.empty())
+      {
+        throw std::invalid_argument(
+            "There is no need to provide an ID (such as client, object, or resource ID) if you are "
+            "using system assigned managed identity.");
+      }
+    }
+
+    /**
+     * @brief Gets the identifier for a user-assigned managed identity.
+     *
+     * @remark In the case of system-assigned managed identity, this will return an empty string.
+     */
+    std::string GetId() const { return m_id; }
```

Comment:
This is a case where returning a const& makes sense - it saves an allocation for most callers.